lib/SILGen/SILGenDestructor.cpp

```diff
@@ -219,6 +221,146 @@ void SILGenFunction::destroyClassMember(SILLocation cleanupLoc,
+    if (Ty && Ty->getCanonicalType() == selfTy->getCanonicalType()) {
+      result.insert(vd);
+    }
+  }
+
+  // NOTE: Right now we only optimize linear recursion, so if there is more
+  // than one stored property of the same type, clear out the set and don't
+  // perform any recursion optimization.
+  if (result.size() > 1) {
+    result.clear();
+  }
+}
+
+void SILGenFunction::emitRecursiveChainDestruction(ManagedValue selfValue,
+                                                   ClassDecl *cd,
+                                                   VarDecl *recursiveLink,
+                                                   CleanupLocation cleanupLoc) {
+  auto selfTy = F.mapTypeIntoContext(cd->getDeclaredInterfaceType());
+
+  auto selfTyLowered = getTypeLowering(selfTy).getLoweredType();
+
+  SILBasicBlock *cleanBB = createBasicBlock();
+  SILBasicBlock *noneBB = createBasicBlock();
+  SILBasicBlock *notUniqueBB = createBasicBlock();
+  SILBasicBlock *uniqueBB = createBasicBlock();
+  SILBasicBlock *someBB = createBasicBlock();
+  SILBasicBlock *loopBB = createBasicBlock();
+
+  // var iter = self.link
+  // self.link = nil
+  auto Ty = getTypeLowering(F.mapTypeIntoContext(recursiveLink->getInterfaceType())).getLoweredType();
+  auto optionalNone = B.createOptionalNone(cleanupLoc, Ty);
+  SILValue varAddr =
+    B.createRefElementAddr(cleanupLoc, selfValue.getValue(), recursiveLink,
+                           Ty.getAddressType());
+  auto *iterAddr = B.createAllocStack(cleanupLoc, Ty);
+  SILValue addr = B.createBeginAccess(
+    cleanupLoc, varAddr, SILAccessKind::Modify, SILAccessEnforcement::Static,
+    true /*noNestedConflict*/, false /*fromBuiltin*/);
+  SILValue iter = B.createLoad(cleanupLoc, addr, LoadOwnershipQualifier::Take);
+  B.createStore(cleanupLoc, optionalNone, addr, StoreOwnershipQualifier::Init);
+  B.createEndAccess(cleanupLoc, addr, false /*is aborting*/);
+  B.createStore(cleanupLoc, iter, iterAddr, StoreOwnershipQualifier::Init);
+
+  B.createBranch(cleanupLoc, loopBB);
+
+  // while iter != nil {
+  {
+    B.emitBlock(loopBB);
+    auto iterBorrow =
+        ManagedValue::forUnmanaged(iterAddr).borrow(*this, cleanupLoc);
+    SwitchEnumBuilder switchBuilder(B, cleanupLoc, iterBorrow);
+    switchBuilder.addOptionalSomeCase(someBB);
+    switchBuilder.addOptionalNoneCase(noneBB);
+    std::move(switchBuilder).emit();
+  }
+
+  // if isKnownUniquelyReferenced(&iter) {
+  {
+    B.emitBlock(someBB);
+    auto isUnique = B.createIsUnique(cleanupLoc, iterAddr);
+    B.createCondBranch(cleanupLoc, isUnique, uniqueBB, notUniqueBB);
+  }
+
+  // we have a uniquely referenced link, so we need to deinit
+  {
+    B.emitBlock(uniqueBB);
+
+    // NOTE: We increment the ref count of the tail instead of unlinking it,
+    //       because custom deinit implementations of subclasses may access
+    //       it and it would be semantically wrong to unset it before that.
```

Comment:
I think you can get rid of this part of the comment. I think the deinit of tier comment below is sufficient.